feat(eslint-plugin): add no-unsafe-declaration-merging

yeonjuan opened the pull request:
<!--
👋 Hi, thanks for sending a PR to typescript-eslint! 💖
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

- [x] Addresses an existing open issue: fixes #5275
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/no-unsafe-declaration-merging.test.ts:
```diff
@@ -0,0 +1,71 @@
+      ],
+    },
+  ],
+});
```
These tests are quite good for the main cases, but don't cover very many edge cases. How about some that cover...

* Declaring a class within odd places:
    * Inside a function declaration and IIFE
    * Inside a class extends clause, like `class Apple extends class Banana {}`
* Inside a `declare global`

JoshuaKGoldberg reviewed (changes requested):
Looks great, thanks for sending this in! Mostly requesting a bit more test coverage.

yeonjuan commented:
@JoshuaKGoldberg Thanks for the reviews. I fixed it.

And, I changed the reported node with the Identifier node, because reporting the whole class or interface declaration can overlap other errors inside the declaration body.

```ts
// before
context.report({
            node: node.parent,
            messageId: 'unsafeMerging',
          });
 // current
 context.report({
            node: node.parent,
            messageId: 'unsafeMerging',
          });       
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-unsafe-declaration-merging.md:
```diff
@@ -0,0 +1,54 @@
+TypeScript's "declaration merging" supports merging separate declarations with the same name.
+
+Declaration merging between classes and interfaces is unsafe.
+TypeScript compiler doesn't check whether properties are initialized, which can cause lead to TypeScript not detecting code that will cause runtime errors.
```
Nit: just one more grammar thing...

```suggestion
The TypeScript compiler doesn't check whether properties are initialized, which can cause lead to TypeScript not detecting code that will cause runtime errors.
```

JoshuaKGoldberg commented on packages/eslint-plugin/docs/rules/no-unsafe-declaration-merging.md:
````diff
@@ -0,0 +1,54 @@
+
+```ts
+interface Foo {}
+class Bar extends Foo {}
````
`extends` fails type checking here, I think you meant this?

```suggestion
class Bar implements Foo {}
```

JoshuaKGoldberg reviewed (dismissed):
Fantastic, thanks for sending this in @yeonjuan! 🙌

bradzacher commented on packages/eslint-plugin/docs/rules/no-unsafe-declaration-merging.md:
```diff
@@ -0,0 +1,54 @@
+TypeScript's "declaration merging" supports merging separate declarations with the same name.
+
+Declaration merging between classes and interfaces is unsafe.
+The TypeScript compiler doesn't check whether properties are initialized, which can cause lead to TypeScript not detecting code that will cause runtime errors.
```
oops typo!

```suggestion
The TypeScript compiler doesn't check whether properties are initialized, which can lead to TypeScript not detecting code that will cause runtime errors.
```

bradzacher commented on packages/eslint-plugin/src/rules/no-unsafe-declaration-merging.ts:
```diff
@@ -0,0 +1,52 @@
+    docs: {
+      description: 'Disallow unsafe declaration merging',
+      recommended: 'strict',
+      requiresTypeChecking: true,
```
We should actually be able to do this without type info and just with the scope analysis APIs.

![image](https://user-images.githubusercontent.com/7462525/196681607-6c635160-a47b-4746-842c-636d303be9b4.png)
[playground](https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&showAST=scope&code=JYOwLgpgTgZghgYwgAgGIHt3IN7IL4CwAUAgDZwDOFamO+QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

You can see that if the names clash, then we can iterate through the `variable.defs` array and do a similar comparison as you've done here - `variable.defs.some(def => def.node.type === unsafeType)` where
`unsafeType == AST_NODE_TYPES.TSInterfaceDeclaration | AST_NODE_TYPES.ClassDeclaration`

JoshuaKGoldberg replied:
Ah, I should finally get familiar with the scope manager 😄. #5854

JoshuaKGoldberg replied:
@yeonjuan do you have time to tackle this soon? I can if not, no worries 😄

yeonjuan replied:
Hi @bradzacher, @JoshuaKGoldberg  I can fix this soon. I'll make a PR for it. (including [fixing typo](https://github.com/typescript-eslint/typescript-eslint/pull/5840#discussion_r999326647))
 Thanks!
